atc/db/job_factory.go

```diff
@@ -30,6 +35,148 @@ func NewJobFactory(conn Conn, lockFactory lock.LockFactory) JobFactory {
 	}
 }
 
+type SchedulerJobs []SchedulerJob
+
+type SchedulerJob struct {
+	Job
+	Resources     SchedulerResources
+	ResourceTypes atc.VersionedResourceTypes
+}
+
+type SchedulerResources []SchedulerResource
+
+type SchedulerResource struct {
+	Name   string
+	Type   string
+	Source atc.Source
+	Tags   atc.Tags
+}
+
+func (resources SchedulerResources) Lookup(name string) (SchedulerResource, bool) {
+	for _, resource := range resources {
+		if resource.Name == name {
+			return resource, true
+		}
+	}
+
+	return SchedulerResource{}, false
+}
+
+func (j *jobFactory) JobsToSchedule() (SchedulerJobs, error) {
+	tx, err := j.conn.Begin()
+	if err != nil {
+		return nil, err
+	}
+
+	defer tx.Rollback()
+
+	rows, err := jobsQuery.
+		Where(sq.Expr("j.schedule_requested > j.last_scheduled")).
+		Where(sq.Eq{
+			"j.active": true,
+			"j.paused": false,
+			"p.paused": false,
+		}).
+		RunWith(tx).
+		Query()
+	if err != nil {
+		return nil, err
+	}
+
+	jobs, err := scanJobs(j.conn, j.lockFactory, rows)
+	if err != nil {
+		return nil, err
+	}
+
+	var schedulerJobs SchedulerJobs
+	pipelineResourceTypes := make(map[int]ResourceTypes)
+	for _, job := range jobs {
+		rows, err := psql.Select("r.name", "r.type", "r.config", "r.nonce").
```

Comment:
Is looking up the resources for each job better than doing the top level resources query?

Reply:
@pivotal-jwinters You mean like using the `job_inputs` table to find what resources are for that job, check if they use any custom resource types and then grab out all the resource types that are used by that job?

Reply:
I just mean we used to grab all the resources for a pipeline up front:
https://github.com/concourse/concourse/blob/master/atc/scheduler/runner.go#L80

But here we are making many queries to fetch resources per job. I was just wondering if you found that this was some way better?

But now that you mention resource types, It also seems weird because the only thing that needs the resource types is the the build starter (is that true?). But now we're fetching the resource types super early in the scheduler (which doesn't actually need them)? To me, the scheduler and build starter should be looked at as two different components but this seems to couple their behaviour.

Reply:
Ohh sorry I thought you were talking about the query for resource types. If you mean better as in faster or less duplicate work, I'm *pretty* sure doing it once per pipeline and then just passing all of them to each job is technically faster and less duplicate work because it is a single bigger query rather than multiple smaller queries that are probably grabbing the same resources.

I think what we were trying to achieve is to have the scheduler be completely job focused and no longer needs to care about pipelines. But I could definitely be convinced that the performance overhead is not worth it.

Reply:
Oh i see. Yeah making it entirely job focused makes a lot of sense to me. What if we just query for ALL resources up front? Similar to what the lidar scanner does?

Reply:
Hmm if we did that we would still need to figure out which resources go with which jobs, so we would need to query to the `job_inputs` table for each job right?

Reply:
What if we do that whole job_inputs joining resources query upfront for all jobs?

Reply:
Just a thought too, now that we only schedule jobs that need to be scheduled it might not be a problem doing a query for the resources for each job? I'm just worried that combining it all into one query might make the code more complicated especially if its for optimizing something that might not be a problem.

Reply:
Hmm yeah good point. okey-dokey